### DISCUSSION TITLE
Rewrite values in the query

### DISCUSSION THREAD

**Discussion by romankydybets** [esnet/gdg#121] [Q&A (Help Section)] [answered]
During the process for import/export dashboard through the environments, sometime i do need rewrite some parts inside the query. Unfortunately they couldn't be covered by regular expressions, 
Is there any chance to do it with your tool?

**Comment by csg33k** [comment]
The Regex is intended primarily for determining which datasource to use and filtering purposes.  There's nothing in the too currently that supports rewrites.  I'd like to add support for plugins though it was initially intended mainly to provide more flexibility for input and output destination.  

I wrote this bit of code a few years back not sure how useful it would be but it deals with manipulating the data a bit, query and such.  Some of it is very custom to netsage project, other might be reusable.   

https://github.com/netsage-project/netsage-grafana-configs/tree/master/templates

I will say that regex manipulation and such is really error prone.  If you really want to control how a dashboard looks you might look at: https://grafana.com/blog/2022/12/06/a-complete-guide-to-managing-grafana-as-code-tools-tips-and-tricks/.  

Downside it's a 1 way route.  Aka can't easily import the changes someone made on a dashboard back into code.   GDG and similar tool rely on the API which is great for imports/exports but modifying the generated JSON isn't always very easy to do.

**Comment by csg33k** [comment]
TLDR: I don't mind adding this feature but I will say that this might be something I'd rather support via plugins and might take a while before we're at that point.

**Comment by safaci2000** [comment]
@romankydybets I may revisit this but I imagine we'll need to have some embedded scripting language to support this.

I've used some JS engine with Go, though I need to explore others.  I'd prefer golang but not sure if it has enough adoption to really make sense.  

Anyways, this might be getting revisited.

**Reply by romankydybets** [reply]
Previously we had python which takes the dashboards, modify and storing it.
I just decided to raise the question, because I think that this would quite useful.

**Reply by safaci2000** [reply]
I think it would be an interesting pattern to support.  The problem is there is no real easy way to script this.  The output JSON that Grafana spits is a bit of a mess.  It seems constantly changing the JSON if I manage it in Git even if no changes happened for example.  

We could do something very basic easily enough.

ie.  

- rules:
      field:  "title"
      regex: ".*Some Pattern.*"
      replace: "somereex or string"
      
I think that should be okay and easy enough to do, but I found myself needing to do more complex operations.

My use cases usually tend to be more of ....   in the panels array, find the panel of type "foobar", and in that query, please update the options.content field to be X instead of Y, or replace this sentence with this one.  

Basically you'd need to embed some logic.   So the question is what should GDG support.   This gets a bit away from just here's a tool with a simple config that a wizard will get you 90% of the way there to get started.  Now we have a embedded engine where you're writing code. 

Not even looking at what is / isn't supported in Go.  

My own biases tend to lean towards python / Go.  Lua could be interesting though I'd like to avoid CGO or any C dependencies.  One of my favorite aspects of Go is that it's one static file and you're good to go.  You can take the GDG binary and run through the config wizard and it would basically give you all you need. 

Food for thought I guess.

**Reply by romankydybets** [reply]
i can share the peace of code, which is taking care of this, or we could discuss, on some call.

**Reply by safaci2000** [reply]
Can you add more context on your use case.  I'm going to revisit this request in the next few weeks.  https://github.com/esnet/gdg/issues/168

**Reply by safaci2000** [reply]
@romankydybets  Just an FYI, I have an initial version of this under code review. If you want to take a pass at what GDG can support and see if it satisfies your needs. 

The next release will include gdg-generate which is a template driven tool.  It'll read files under "templates" and you can use it combined with templates.yml to generate any number of dashboards.  It'll create the files in the path of the given context. 

So if context is set to testing with output of test/data/. the dashboards will go into test/data/dashboards/org_id/<name>.json 

The doc will be more polished on released, but let me know if that makes sense?

**Reply by romankydybets** [reply]
Hey @safaci2000, 
Thank you for this feature.
I didn't have a chance to test it, but I am sure that it would help.

**Reply by safaci2000** [reply, marked as answer]
Sure thing. 

Let me know if there's any ambiguity: https://software.es.net/gdg/docs/templating/usage-guide/ though I will say this is a one way workflow.  Template -> dashboards, you can't really pull changes from a live dashboard, pop them into a template and have it update all dashboards again.  (Well, not without dealing with essentially the equivalent of a merge conflict)
